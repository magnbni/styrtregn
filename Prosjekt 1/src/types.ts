interface Day {
    day: Record<string, JSONArray>
}

interface Geo {
    lat: string,
    lon: string
}

interface WeatherData {
    date: string
}

interface Hour {
    temp_c: string,
    time: string,
    condition: Condition
}

interface Condition {
    icon: string
}

type JSONvalue = string | number | string[] | Array<JSONvalue> | Hour;



interface JSONArray extends Array<Hour> { }