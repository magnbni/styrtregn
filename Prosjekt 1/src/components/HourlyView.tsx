import "./hourlyview.css"
import { Properties } from "../types";
import FetchIcon from "../FetchIcon";
import "./Popup.css"
/**
 * This function returns a component that only shows the hourly forecast per
 * given date.
 * 
 * @param props 
 * @param showBoolean: boolean
 * @param metData: Properties
 * @param day: string
 * @returns
 */
function HourlyView(props: { metData: Properties, day: string}) {
    
    const { metData, day} = props;

    const returnTableData = () => {
        return metData.timeseries.filter(e => e.time.slice(0,10) == day.slice(0,10)).map(
            (res) =>
            <>
                <tr key="time">
                    <td>{res.time.slice(11,16)}</td>
                    <td style={{ width: "15%" }}>
                    {res.data.next_1_hours?.summary.symbol_code ? FetchIcon(res.data.next_1_hours?.summary.symbol_code) : <p></p>}
                    </td>
                    <td>{res.data.instant.details.air_temperature} °C</td>
                    <td>{ res.data.next_1_hours?.details.precipitation_amount ? res.data.next_1_hours.details.precipitation_amount : res.data.next_6_hours?.details.precipitation_amount}</td>
                    <td>{res.data.instant.details.wind_speed}</td>
                </tr>
            </>
        )
    }

    
    return (
        <div className="overlay">
            <h2>Været {day.slice(0,10)}</h2>
            <div className="fluid-table">
                <div className="table-container">
                <table className="table2">
                    <thead>
                        <tr>
                            {/* Replace with tableheader function for ease of implentation? */}
                            <td>Klokkeslett</td>
                            <td>Vær</td>
                            <td>Temperatur</td>
                            <td>Nedbør mm</td>
                            <td>Vind m/s</td>
                        </tr>
                    </thead>
                    <tbody>
                        {/* Replace with tablebody function for ease of implementation and cleaner code */}
                        {returnTableData()}
                    </tbody>
                </table>
                </div>
            </div>
        </div>
    );
}



export default HourlyView;