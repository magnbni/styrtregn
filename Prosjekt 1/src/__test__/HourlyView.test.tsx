import { expect, it } from 'vitest'
import HourlyView from '../components/HourlyView'
import { Properties } from '../types';

it('The HourlyView component should output the same HTML with the same props', () => {
    const props: { metData: Properties; day: string } = {
        metData: {
            meta: {
                updated_at: '2021-10-10T09:00:00Z',
                units: {
                    air_pressure_at_sea_level: 'hPa',
                    air_temperature: 'celsius',
                    cloud_area_fraction: '%',
                    precipitation_amount: 'mm',
                    relative_humidity: '%',
                    wind_from_direction: 'degrees',
                    wind_speed: 'm/s',
                },
            },
            timeseries: [
                {
                    time: '2021-10-10T09:00:00Z',
                    data: {
                        instant: {
                            details: {
                                air_pressure_at_sea_level: 1000.6,
                                air_temperature: 8.9,
                                cloud_area_fraction: 100,
                                relative_humidity: 93.4,
                                wind_from_direction: 237.5,
                                wind_speed: 2.4,
                            },
                        },
                        next_1_hours: {
                            summary: {
                                symbol_code: 'rainshowers_day',
                            },
                            details: {
                                precipitation_amount: 0.3,
                            },
                        },
                        next_6_hours: {
                            summary: {
                                symbol_code: 'rainshowers_day',
                            },
                            details: {
                                precipitation_amount: 0.4,
                            },
                        },
                    },
                },
            ],
        },
        day: '2021-10-10'}
    const hourlyView = HourlyView(props)
    expect(hourlyView).toMatchSnapshot()
})