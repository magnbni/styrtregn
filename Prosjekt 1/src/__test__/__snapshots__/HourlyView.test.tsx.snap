// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The HourlyView component should output the same HTML with the same props 1`] = `
<div
  className="overlay"
>
  <h2>
    Været 
    2021-10-10
  </h2>
  <div
    className="fluid-table"
  >
    <div
      className="table-container"
    >
      <table
        className="table2"
      >
        <thead>
          <tr>
            <td>
              Klokkeslett
            </td>
            <td>
              Vær
            </td>
            <td>
              Temperatur
            </td>
            <td>
              Nedbør mm
            </td>
            <td>
              Vind m/s
            </td>
          </tr>
        </thead>
        <tbody>
          <React.Fragment>
            <tr>
              <td>
                09:00
              </td>
              <td
                style={
                  {
                    "width": "15%",
                  }
                }
              >
                <img
                  alt="icon"
                  className="weathericon"
                  src="/project1/src/weathericons-svg/rainshowers_day.svg"
                />
              </td>
              <td>
                8.9
                 °C
              </td>
              <td>
                0.3
              </td>
              <td>
                2.4
              </td>
            </tr>
          </React.Fragment>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
